# Regression Week 4: Ridge Regression (interpretation)

In this notebook, we will run ridge regression multiple times with different L2 penalties to see which one produces the best fit. We will revisit the example of polynomial regression as a means to see the effect of L2 regularization. In particular, we will:
* Use a pre-built implementation of regression (GraphLab Create) to run polynomial regression
* Use matplotlib to visualize polynomial regressions
* Use a pre-built implementation of regression (GraphLab Create) to run polynomial regression, this time with L2 penalty
* Use matplotlib to visualize polynomial regressions under L2 regularization
* Choose best L2 penalty using cross-validation.
* Assess the final fit using test data.

We will continue to use the House data from previous notebooks.  (In the next programming assignment for this module, you will implement your own ridge regression learning algorithm using gradient descent.)

# Fire up graphlab create

In [156]:
import graphlab
import pandas as pd
import numpy as np

# Polynomial regression, revisited

We build on the material from Week 3, where we wrote the function to produce an SFrame with columns containing the powers of a given input. Copy and paste the function `polynomial_sframe` from Week 3:

In [157]:
def polynomial_sframe(feature, degree):
    # assume that degree >= 1
    # initialize the SFrame:
    poly_sframe = pd.DataFrame()
    # and set poly_sframe['power_1'] equal to the passed feature
    poly_sframe['power_1']= feature
    # first check if degree > 1
    if degree > 1:
        # then loop over the remaining degrees:
        # range usually starts at 0 and stops at the endpoint-1. We want it to start at 2 and stop at degree
        for power in range(2, degree+1): 
            # first we'll give the column a name:
            name = 'power_' + str(power)
            # then assign poly_sframe[name] to the appropriate power of feature
            poly_sframe[name]=poly_sframe['power_1'].apply(lambda x: x**power)

    return poly_sframe
    

Let's use matplotlib to visualize what a polynomial regression looks like on the house data.

In [193]:
def polynomial_sframe(feature, degree):
    # assume that degree >= 1
    # initialize the SFrame:
    poly_sframe = graphlab.SFrame()
    # and set poly_sframe['power_1'] equal to the passed feature
    poly_sframe['power_1']= feature
    # first check if degree > 1
    if degree > 1:
        # then loop over the remaining degrees:
        # range usually starts at 0 and stops at the endpoint-1. We want it to start at 2 and stop at degree
        for power in range(2, degree+1): 
            # first we'll give the column a name:
            name = 'power_' + str(power)
            # then assign poly_sframe[name] to the appropriate power of feature
            poly_sframe[name]=poly_sframe['power_1'].apply(lambda x: x**power)

    return poly_sframe

In [158]:
import matplotlib.pyplot as plt
%matplotlib inline

In [159]:

dtype_dict = {'bathrooms':float, 'waterfront':int, 'sqft_above':int, 'sqft_living15':float, 'grade':int, 'yr_renovated':int, 'price':float, 'bedrooms':float, 'zipcode':str, 'long':float, 'sqft_lot15':float, 'sqft_living':float, 'floors':float, 'condition':int, 'lat':float, 'date':str, 'sqft_basement':int, 'yr_built':int, 'id':str, 'sqft_lot':int, 'view':int}

sales = pd.read_csv('kc_house_data.csv', dtype=dtype_dict)
sales = sales.sort_values(['sqft_living','price'])

In [ ]:
sales = graphlab.SFrame('kc_house_data.gl/')

As in Week 3, we will use the sqft_living variable. For plotting purposes (connecting the dots), you'll need to sort by the values of sqft_living. For houses with identical square footage, we break the tie by their prices.

In [160]:
sales = sales.sort(['sqft_living','price'])

AttributeError: 'DataFrame' object has no attribute 'sort'

Let us revisit the 15th-order polynomial model using the 'sqft_living' input. Generate polynomial features up to degree 15 using `polynomial_sframe()` and fit a model with these features. When fitting the model, use an L2 penalty of `1e-5`:

In [6]:
l2_small_penalty = 1e-5

Note: When we have so many features and so few data points, the solution can become highly numerically unstable, which can sometimes lead to strange unpredictable results.  Thus, rather than using no regularization, we will introduce a tiny amount of regularization (`l2_penalty=1e-5`) to make the solution numerically stable.  (In lecture, we discussed the fact that regularization can also help with numerical stability, and here we are seeing a practical example.)

With the L2 penalty specified above, fit the model and print out the learned weights.

Hint: make sure to add 'price' column to the new SFrame before calling `graphlab.linear_regression.create()`. Also, make sure GraphLab Create doesn't create its own validation set by using the option `validation_set=None` in this call.

In [161]:
from sklearn import linear_model
import numpy as np

poly15_data = polynomial_sframe(sales['sqft_living'], 15) # use equivalent of `polynomial_sframe`
model = linear_model.Ridge(alpha=l2_small_penalty, normalize=True)
model.fit(poly15_data, sales['price'])


Ridge(alpha=1e-09, copy_X=True, fit_intercept=True, max_iter=None,
   normalize=True, random_state=None, solver='auto', tol=0.001)

In [12]:
len(model.fit(poly15_data, sales['price']).coef_)
model.fit(poly15_data, sales['price']).coef_

array([ 1.74425503e+02, -7.83644074e-02,  3.73280427e-05, -2.85554311e-09,
       -2.37487557e-13,  9.70529709e-18,  1.77791834e-21,  9.76993158e-26,
        7.78897288e-31, -3.83833667e-34, -4.52940095e-38, -3.16694117e-42,
       -1.24102475e-46,  4.53456467e-51,  1.63425896e-54])

***QUIZ QUESTION:  What's the learned value for the coefficient of feature `power_1`?***

# Observe overfitting

Recall from Week 3 that the polynomial fit of degree 15 changed wildly whenever the data changed. In particular, when we split the sales data into four subsets and fit the model of degree 15, the result came out to be very different for each subset. The model had a *high variance*. We will see in a moment that ridge regression reduces such variance. But first, we must reproduce the experiment we did in Week 3.

First, split the data into split the sales data into four subsets of roughly equal size and call them `set_1`, `set_2`, `set_3`, and `set_4`. Use `.random_split` function and make sure you set `seed=0`. 

In [166]:
set_1 = pd.read_csv('wk3_kc_house_set_1_data.csv', dtype=dtype_dict)
set_2 = pd.read_csv('wk3_kc_house_set_2_data.csv', dtype=dtype_dict)
set_3 = pd.read_csv('wk3_kc_house_set_3_data.csv', dtype=dtype_dict)
set_4 = pd.read_csv('wk3_kc_house_set_4_data.csv', dtype=dtype_dict)

Next, fit a 15th degree polynomial on `set_1`, `set_2`, `set_3`, and `set_4`, using 'sqft_living' to predict prices. Print the weights and make a plot of the resulting model.

Hint: When calling `graphlab.linear_regression.create()`, use the same L2 penalty as before (i.e. `l2_small_penalty`).  Also, make sure GraphLab Create doesn't create its own validation set by using the option `validation_set = None` in this call.

In [20]:
l2_small_penalty=1e-9

In [167]:
poly_1 = polynomial_sframe(set_1['sqft_living'], 15)
poly_2 = polynomial_sframe(set_2['sqft_living'], 15)
poly_3 = polynomial_sframe(set_3['sqft_living'], 15)
poly_4 = polynomial_sframe(set_4['sqft_living'], 15)# use equivalent of `polynomial_sframe`
model = linear_model.Ridge(alpha=l2_small_penalty, normalize=True)
model1= model.fit(poly_1, set_1['price'])
# model2= model.fit(poly_2, set_2['price'])
# model3= model.fit(poly_3, set_3['price'])
# model4= model.fit(poly_4, set_4['price'])
model1.coef_[0]

544.6693903291846

In [168]:
model2= model.fit(poly_2, set_2['price'])
model2.coef_[0]

859.3626475978917

In [169]:
model3= model.fit(poly_3, set_3['price'])
model3.coef_[0]

-755.3959047401938

In [170]:
model4= model.fit(poly_4, set_4['price'])

model4.coef_[0]

1119.4457483861213

In [81]:
#model4.coef_[0]
model4.predict(poly_4)

array([ 525203.86613591,  525394.22999686,  525418.46787149, ...,
        811422.9537882 ,  821997.62146415, 1065880.04623581])

The four curves should differ from one another a lot, as should the coefficients you learned.

***QUIZ QUESTION:  For the models learned in each of these training sets, what are the smallest and largest values you learned for the coefficient of feature `power_1`?***  (For the purpose of answering this question, negative numbers are considered "smaller" than positive numbers. So -5 is smaller than -3, and -3 is smaller than 5 and so forth.)

# Ridge regression comes to rescue

Generally, whenever we see weights change so much in response to change in data, we believe the variance of our estimate to be large. Ridge regression aims to address this issue by penalizing "large" weights. (Weights of `model15` looked quite small, but they are not that small because 'sqft_living' input is in the order of thousands.)

With the argument `l2_penalty=1e5`, fit a 15th-order polynomial model on `set_1`, `set_2`, `set_3`, and `set_4`. Other than the change in the `l2_penalty` parameter, the code should be the same as the experiment above. Also, make sure GraphLab Create doesn't create its own validation set by using the option `validation_set = None` in this call.

In [53]:
l2_large_penalty=1.23e2
model = linear_model.Ridge(alpha=l2_large_penalty, normalize=True)
model1= model.fit(poly_1, set_1['price'])
model1.coef_[0]

2.328068029579326

In [54]:
model2= model.fit(poly_2, set_2['price'])
model2.coef_[0]

2.097569027778555

In [55]:
model3= model.fit(poly_3, set_3['price'])
model3.coef_[0]

2.289062581189205

In [56]:
model4= model.fit(poly_4, set_4['price'])
model4.coef_[0]

2.0859619409193066

These curves should vary a lot less, now that you applied a high degree of regularization.

***QUIZ QUESTION:  For the models learned with the high level of regularization in each of these training sets, what are the smallest and largest values you learned for the coefficient of feature `power_1`?*** (For the purpose of answering this question, negative numbers are considered "smaller" than positive numbers. So -5 is smaller than -3, and -3 is smaller than 5 and so forth.)

# Selecting an L2 penalty via cross-validation

Just like the polynomial degree, the L2 penalty is a "magic" parameter we need to select. We could use the validation set approach as we did in the last module, but that approach has a major disadvantage: it leaves fewer observations available for training. **Cross-validation** seeks to overcome this issue by using all of the training set in a smart way.

We will implement a kind of cross-validation called **k-fold cross-validation**. The method gets its name because it involves dividing the training set into k segments of roughtly equal size. Similar to the validation set method, we measure the validation error with one of the segments designated as the validation set. The major difference is that we repeat the process k times as follows:

Set aside segment 0 as the validation set, and fit a model on rest of data, and evalutate it on this validation set<br>
Set aside segment 1 as the validation set, and fit a model on rest of data, and evalutate it on this validation set<br>
...<br>
Set aside segment k-1 as the validation set, and fit a model on rest of data, and evalutate it on this validation set

After this process, we compute the average of the k validation errors, and use it as an estimate of the generalization error. Notice that  all observations are used for both training and validation, as we iterate over segments of data. 

To estimate the generalization error well, it is crucial to shuffle the training data before dividing them into segments. GraphLab Create has a utility function for shuffling a given SFrame. We reserve 10% of the data as the test set and shuffle the remainder. (Make sure to use `seed=1` to get consistent answer.)

In [162]:
train_valid_shuffled = pd.read_csv('wk3_kc_house_train_valid_shuffled.csv', dtype=dtype_dict)
test = pd.read_csv('wk3_kc_house_test_data.csv', dtype=dtype_dict)

Once the data is shuffled, we divide it into equal segments. Each segment should receive `n/k` elements, where `n` is the number of observations in the training set and `k` is the number of segments. Since the segment 0 starts at index 0 and contains `n/k` elements, it ends at index `(n/k)-1`. The segment 1 starts where the segment 0 left off, at index `(n/k)`. With `n/k` elements, the segment 1 ends at index `(n*2/k)-1`. Continuing in this fashion, we deduce that the segment `i` starts at index `(n*i/k)` and ends at `(n*(i+1)/k)-1`.

With this pattern in mind, we write a short loop that prints the starting and ending indices of each segment, just to make sure you are getting the splits right.

In [58]:
n = len(train_valid_shuffled)
k = 10 # 10-fold cross-validation

for i in xrange(k):
    start = (n*i)/k
    end = (n*(i+1))/k-1
    print i, (start, end)

0 (0, 1938)
1 (1939, 3878)
2 (3879, 5817)
3 (5818, 7757)
4 (7758, 9697)
5 (9698, 11636)
6 (11637, 13576)
7 (13577, 15515)
8 (15516, 17455)
9 (17456, 19395)


Let us familiarize ourselves with array slicing with SFrame. To extract a continuous slice from an SFrame, use colon in square brackets. For instance, the following cell extracts rows 0 to 9 of `train_valid_shuffled`. Notice that the first index (0) is included in the slice but the last index (10) is omitted.

In [65]:
train_valid_shuffled[0:10] # rows 0 to 9

id             date     price  bedrooms  bathrooms  sqft_living  \
0  2780400035  20140505T000000  665000.0       4.0       2.50       2800.0   
1  1703050500  20150321T000000  645000.0       3.0       2.50       2490.0   
2  5700002325  20140605T000000  640000.0       3.0       1.75       2340.0   
3  0475000510  20141118T000000  594000.0       3.0       1.00       1320.0   
4  0844001052  20150128T000000  365000.0       4.0       2.50       1904.0   
5  2781280290  20150427T000000  305000.0       3.0       2.50       1610.0   
6  2214800630  20141105T000000  239950.0       3.0       2.25       1560.0   
7  2114700540  20141021T000000  366000.0       3.0       2.50       1320.0   
8  2596400050  20140730T000000  375000.0       3.0       1.00       1960.0   
9  4140900050  20150126T000000  440000.0       4.0       1.75       2180.0   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5900     1.0           0     0  ...      8        1660           1140   
1      5978     2.0           0     0  ...      9        2490              0   
2      4206     1.0           0     0  ...      7        1170           1170   
3      5000     1.0           0     0  ...      7        1090            230   
4      8200     2.0           0     0  ...      7        1904              0   
5      3516     2.0           0     0  ...      8        1610              0   
6      8280     2.0           0     0  ...      7        1560              0   
7      4320     1.0           0     0  ...      6         660            660   
8      7955     1.0           0     0  ...      7        1260            700   
9     10200     1.0           0     2  ...      8        2000            180   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1963             0    98115  47.6809 -122.286         2580.0   
1      2003             0    98074  47.6298 -122.022         2710.0   
2      1917             0    98144  47.5759 -122.288         1360.0   
3      1920             0    98107  47.6674 -122.365         1700.0   
4      1999             0    98010  47.3107 -122.001         1560.0   
5      2006             0    98055  47.4491 -122.188         1610.0   
6      1979             0    98001  47.3393 -122.259         1920.0   
7      1918             0    98106  47.5327 -122.347         1190.0   
8      1963             0    98177  47.7641 -122.364         1850.0   
9      1966             0    98028  47.7638 -122.270         2590.0   

   sqft_lot15  
0      5900.0  
1      6629.0  
2      4725.0  
3      5000.0  
4     12426.0  
5      3056.0  
6      8120.0  
7      4200.0  
8      8219.0  
9     10445.0  

[10 rows x 21 columns]

Now let us extract individual segments with array slicing. Consider the scenario where we group the houses in the `train_valid_shuffled` dataframe into k=10 segments of roughly equal size, with starting and ending indices computed as above.
Extract the fourth segment (segment 3) and assign it to a variable called `validation4`.

In [70]:
validation4= train_valid_shuffled[5818:7758]#, 7757]
validation4['price'].mean()

536234.1

To verify that we have the right elements extracted, run the following cell, which computes the average price of the fourth segment. When rounded to nearest whole number, the average should be $536,234.

In [71]:
print int(round(validation4['price'].mean(), 0))

536234


After designating one of the k segments as the validation set, we train a model using the rest of the data. To choose the remainder, we slice (0:start) and (end+1:n) of the data and paste them together. SFrame has `append()` method that pastes together two disjoint sets of rows originating from a common dataset. For instance, the following cell pastes together the first and last two rows of the `train_valid_shuffled` dataframe.

In [72]:
n = len(train_valid_shuffled)
first_two = train_valid_shuffled[0:2]
last_two = train_valid_shuffled[n-2:n]
print first_two.append(last_two)

               id             date      price  bedrooms  bathrooms  \
0      2780400035  20140505T000000   665000.0       4.0       2.50   
1      1703050500  20150321T000000   645000.0       3.0       2.50   
19394  4139480190  20140916T000000  1153000.0       3.0       3.25   
19395  7237300290  20150326T000000   338000.0       5.0       2.50   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
0           2800.0      5900     1.0           0     0  ...      8   
1           2490.0      5978     2.0           0     0  ...      9   
19394       3780.0     10623     1.0           0     1  ...     11   
19395       2400.0      4496     2.0           0     0  ...      7   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
0            1660           1140      1963             0    98115  47.6809   
1            2490              0      2003             0    98074  47.6298   
19394        2650           1130      1999             0    9800

Extract the remainder of the data after *excluding* fourth segment (segment 3) and assign the subset to `train4`.

In [77]:
n=len(train_valid_shuffled)
start = (n*i)/10
end = (n*(i+1))/10
i=3
train_valid_shuffled[start:end+1]
train4=train_valid_shuffled[0:start].append(train_valid_shuffled[end+1:n])

To verify that we have the right elements extracted, run the following cell, which computes the average price of the data with fourth segment excluded. When rounded to nearest whole number, the average should be $539,450.

In [78]:
print int(round(train4['price'].mean(), 0))

539459


Now we are ready to implement k-fold cross-validation. Write a function that computes k validation errors by designating each of the k segments as the validation set. It accepts as parameters (i) `k`, (ii) `l2_penalty`, (iii) dataframe, (iv) name of output column (e.g. `price`) and (v) list of feature names. The function returns the average validation error using k segments as validation sets.

* For each i in [0, 1, ..., k-1]:
  * Compute starting and ending indices of segment i and call 'start' and 'end'
  * Form validation set by taking a slice (start:end+1) from the data.
  * Form training set by appending slice (end+1:n) to the end of slice (0:start).
  * Train a linear model using training set just formed, with a given l2_penalty
  * Compute validation error using validation set just formed

In [207]:
def k_fold_cross_validation(k, l2_penalty, data, output_name,features_list ):
    n=len(data)
    arr=[]
    for i in xrange(k):
        start = (n*i)/k
        end = (n*(i+1))/k-1
        validation=data[start:end+1]
        train=data[0:start].append(data[end+1:n])
        model= graphlab.linear_regression.create(train,
                                                  target = output_name,
                                                  features = features_list,
                                                  l2_penalty = l2_penalty,
                                                  l1_penalty=0.,
                                                  validation_set=None,verbose=False)
        #fit= model.fit(train,train[output_name])
        predictions= model.predict(validation)
        res = validation[output_name]-predictions
        #print res
        ressquare=np.array(res**2)
        rss = np.sum(ressquare)
        arr.append(rss) 
    print arr
    return np.mean(arr)
     

In [197]:
poly_data = polynomial_sframe(train_valid_shuffled['sqft_living'], 15)
features    = poly_data.column_names()
poly_data['price']=train_valid_shuffled['price']


In [ ]:
arr=[]
for i in np.logspace(0,7, num=13):   
    error= k_fold_cross_validation(10,i,poly_data,'price',features)
    arr.append(error)
print arr
    

[109025115900198.12, 110903080342621.05, 96246215827960.88, 123038456689825.44, 111508676405397.64, 6363115271018008.0, 134426237326525.6, 128635527488257.66, 113853555884706.16, 207504338276827.9]
[109570764371210.69, 111143725941045.6, 95885522917242.0, 122917462854822.19, 111607983804467.31, 5556845302067829.0, 135087403235433.34, 128884729590541.39, 114147444581044.22, 174232664742276.56]
[109878349929316.66, 111316651676147.97, 95716304553165.36, 122779193988261.62, 111707643644826.58, 3120758271324173.5, 135417640388026.81, 129183015654472.86, 114428187282530.78, 145898942940643.34]
[109780799439700.38, 111381359413299.06, 95809033214874.25, 122701994073140.83, 111923867708342.12, 1049577221180547.0, 135245711219626.25, 129639199334913.02, 114794512143024.12, 131637603340302.55]
[109598765412324.5, 111475021479847.25, 96090934437002.19, 122647834414450.56, 112246290076333.84, 219892693261992.38, 134974366374079.28, 130316428075414.47, 115225653602917.89, 129889265907912.69]
[1098

In [208]:
k_fold_cross_validation(10,100,poly_data,'price',features)

[109689846033036.06, 111402594212698.55, 95909597386407.3, 122670967175742.47, 112047275830079.19, 547396132163725.94, 135115242675809.6, 129889204865144.22, 114973388265685.55, 129995245338459.81]


160908949394678.88

***QUIZ QUESTIONS:  What is the best value for the L2 penalty according to 10-fold validation?***

You may find it useful to plot the k-fold cross-validation errors you have obtained to better understand the behavior of the method.  

In [ ]:
# Plot the l2_penalty values in the x axis and the cross-validation error in the y axis.
# Using plt.xscale('log') will make your plot more intuitive.
plt.xscale('log')
plt.plot(np.logspace(0, 7, num=13),arr,'.')    

Once you found the best value for the L2 penalty using cross-validation, it is important to retrain a final model on all of the training data using this value of `l2_penalty`. This way, your final model will be trained on the entire dataset.

***QUIZ QUESTION: Using the best L2 penalty found above, train a model using all training data. What is the RSS on the TEST data of the model you learn with this L2 penalty? ***

In [149]:
poly=polynomial_sframe(train_valid_shuffled['sqft_living'], 15)
bestmodel= linear_model.Ridge(alpha=100, normalize=True)
fit= bestmodel.fit(poly,train_valid_shuffled['price'])
poly_test=polynomial_sframe(test['sqft_living'], 15)
predictions= fit.predict(poly_test)
res = test['price']-predictions
error = sum(res*res)


In [150]:
np.log10(error), error

(14.442201282031792, 276822433432558.9)

In [128]:
1.00000000e+01

10.0